domain/secretbackend/state/state.go

```diff
@@ -0,0 +1,562 @@
+		return nil, errors.Trace(err)
+	}
+	var m Model
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		return tx.Query(ctx, stmt, sqlair.M{"uuid": uuid}).Get(&m)
+	})
+	if errors.Is(err, sql.ErrNoRows) {
+		return nil, fmt.Errorf("%w model %q%w", errors.NotFound, uuid, errors.Hide(err))
+	}
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if !m.Type.IsValid() {
+		// This should never happen.
+		return nil, fmt.Errorf("invalid model type for model %q", uuid)
+	}
+	return &coremodel.Model{
+		UUID:      coremodel.UUID(m.UUID),
+		Name:      m.Name,
+		ModelType: m.Type,
+	}, nil
+}
+
+// CreateSecretBackend adds a new secret backend.
+func (s *State) CreateSecretBackend(ctx context.Context, backend secretbackend.CreateSecretBackendParams) (string, error) {
+	db, err := s.DB()
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+
+	backendStmt, err := sqlair.Prepare(`
+INSERT INTO secret_backend
+(uuid, name, backend_type, token_rotate_interval) VALUES
+(
+    $SecretBackend.uuid,
+    $SecretBackend.name,
+    $SecretBackend.backend_type,
+    $SecretBackend.token_rotate_interval
+)`, SecretBackend{})
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+	rotateStmt, err := sqlair.Prepare(`
+INSERT INTO secret_backend_rotation
+(backend_uuid, next_rotation_time) VALUES
+(
+    $SecretBackendRotation.backend_uuid,
+    $SecretBackendRotation.next_rotation_time
+)`, SecretBackendRotation{})
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+	configStmt, err := sqlair.Prepare(`
+INSERT INTO secret_backend_config
+(backend_uuid, name, content) VALUES
+(
+    $SecretBackendConfig.backend_uuid,
+    $SecretBackendConfig.name,
+    $SecretBackendConfig.content
+)`, SecretBackendConfig{})
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		sb := SecretBackend{
+			UUID:        backend.ID,
+			Name:        backend.Name,
+			BackendType: backend.BackendType,
+		}
+		if backend.TokenRotateInterval != nil {
+			sb.TokenRotateInterval = domain.NullableDuration{Duration: *backend.TokenRotateInterval, Valid: true}
+		}
+		err := tx.Query(ctx, backendStmt, sb).Run()
+		if err != nil {
+			return fmt.Errorf("cannot insert secret backend %q: %w", backend.Name, err)
+		}
+		if backend.NextRotateTime != nil && !backend.NextRotateTime.IsZero() {
+			err = tx.Query(ctx, rotateStmt,
+				SecretBackendRotation{
+					BackendUUID:      backend.ID,
+					NextRotationTime: sql.NullTime{Time: *backend.NextRotateTime, Valid: true},
+				},
+			).Run()
+			if err != nil {
+				return fmt.Errorf("cannot insert secret backend rotation for %q: %w", backend.Name, err)
+			}
+		}
+		for k, v := range backend.Config {
+			err = tx.Query(ctx, configStmt,
+				SecretBackendConfig{
+					BackendUUID: backend.ID,
+					Name:        k,
+					Content:     v.(string),
+				},
+			).Run()
+			if err != nil {
+				return handleSecretBackendConfigError(err, backend.Name)
+			}
+		}
```

Comment:
This should be in a separate func where we also handle updates. See storage pool attributes for example.